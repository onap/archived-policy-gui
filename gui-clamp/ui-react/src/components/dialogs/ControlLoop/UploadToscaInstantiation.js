/*
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2021 Nordix Foundation.
 *  ================================================================================
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  SPDX-License-Identifier: Apache-2.0
 *  ============LICENSE_END=========================================================
 */

import React, { useState } from "react";
import Modal from "react-bootstrap/Modal";
import Button from "react-bootstrap/Button";
import styled from "styled-components";
import Row from "react-bootstrap/Row";
import Form from "react-bootstrap/Form";
import UploadToscaInstantiationFile from "./UploadToscaInstantiationFile";
import jsYaml from "js-yaml";
import Alert from "react-bootstrap/Alert";

const ModalStyled = styled(Modal)`
  background-color: transparent;
`
const StyledMessagesDiv = styled.div`
  overflow: auto;
  min-width: 100%;
  max-height: 300px;
  padding: 5px 5px 0px 5px;
  text-align: center;
`

const UploadToscaInstantiation = (props) => {
  const [show, setShow] = useState(true);
  const [fileIsSelected, setFileIsSelected] = useState(false);
  const [selectedFile, setSelectedFile] = useState();
  const [jsonObject, setJsonObject] = useState([]);
  const [alertMessages, setAlertMessages] = useState();

  const fileUploadHandler = (event) => {
    event.preventDefault();
    console.log('fileUploadHandler called');

    const file = event.currentTarget.files[0];

    if (file !== undefined) {
      console.log('fileDefined called');
      setSelectedFile(file);
      setFileIsSelected(true);

      const fileReader = new FileReader();
      fileReader.onload = () => {
        const jsonFile = jsYaml.load(fileReader.result, 'utf8')
        setJsonObject(jsonFile)
      }

      fileReader.readAsText(file);
    } else {
      console.log('fileUndefined called');
    }
  }

  const onResponseReceivedHandler = async (response) => {
    console.log('onResponseReceivedHandler called');

    if (await response.ok) {
      setAlertMessages(<Alert variant="success">
        <Alert.Heading>Upload Success</Alert.Heading>
        <p>Tosca Instantiation from { selectedFile.name } was Successfully Uploaded</p>
        <hr/>
        <p>Type: { selectedFile.type }</p><p>Size: { +selectedFile.size / 1000 }Kb</p>
      </Alert>);
    } else {
      setAlertMessages(<Alert variant="danger">
        <Alert.Heading>Upload Failure</Alert.Heading>
        <p>Tosca Instantiation from { selectedFile.name } failed to upload</p>
        <p>Status code: { await response.status }: { response.statusText }</p>
        <p>Response Text: { await response.text() }</p>
        <hr/>
        <p>Type: { selectedFile.type }</p><p>Size: { +selectedFile.size / 1000 }Kb</p>
      </Alert>);
    }
  }

  const handleClose = () => {
    console.log('handleClose called');
    setShow(false);
    props.history.push('/');
  }

  return (
    <ModalStyled size="lg" show={ show } onHide={ handleClose } backdrop="static" keyboard={ false }>
      <Modal.Header closeButton>
        <Modal.Title>Tosca Control Loop - Create Instantiation</Modal.Title>
      </Modal.Header>
      <div style={ { padding: '5px 5px 0px 5px' } }>
        <Modal.Body>
          <Form style={ { paddingTop: '20px' } }>
            <Form.Group as={ Row }>
              <Form.File
                type="file"
                className="custom-file-label"
                id="inputGroupFile01"
                onChange={ fileUploadHandler }
                custom
                accept=".yaml,.yml,.json"
                label={ fileIsSelected ? selectedFile.name : 'Please select a file' }></Form.File>
              <UploadToscaInstantiationFile
                jsonObject={jsonObject}
                onResponseReceived={onResponseReceivedHandler}/>
              <Form.Text>Only .yaml, .yml and .json files are supported</Form.Text>
            </Form.Group>
            <Form.Group as={ Row }>
              <StyledMessagesDiv>
                { alertMessages }
              </StyledMessagesDiv>
            </Form.Group>
          </Form>
        </Modal.Body>
      </div>
      <Modal.Footer>
        <Button variant="secondary" type="null" onClick={ handleClose }>Close</Button>
      </Modal.Footer>
    </ModalStyled>
  )
}

export default UploadToscaInstantiation;
