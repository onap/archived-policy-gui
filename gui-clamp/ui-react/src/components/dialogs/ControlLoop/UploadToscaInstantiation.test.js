/*
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2021 Nordix Foundation.
 *  ================================================================================
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  SPDX-License-Identifier: Apache-2.0
 *  ============LICENSE_END=========================================================
 */

import { mount, shallow } from "enzyme";
import React from "react";
import toJson from "enzyme-to-json";
import UploadToscaInstantiation from "./UploadToscaInstantiation";
import { createMemoryHistory } from "history";
import { act } from "react-dom/test-utils";

describe("Verify UploadToscaInstantiation", () => {
  const fs = require("fs");
  const jsonFile = fs.readFileSync("src/components/dialogs/Policy/toscaInstantiationData.test.json");
  const file = new Blob([jsonFile], { type: 'file' });

  it("renders without crashing", () => {
    shallow(<UploadToscaInstantiation/>);
  });

  it("renders correctly", () => {
    const tree = shallow(<UploadToscaInstantiation/>);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should have a Button element', () => {
    const container = shallow(<UploadToscaInstantiation/>)
    expect(container.find('Button').length).toEqual(1);
  });

  it('handleClose called when bottom button clicked', () => {
    const history = createMemoryHistory();
    const component = mount(<UploadToscaInstantiation history={ history }/>)
    const logSpy = jest.spyOn(console, 'log');


    act(() => {
      component.find('[variant="secondary"]').simulate('click');
      expect(logSpy).toHaveBeenCalledWith('handleClose called');
    });
  });

  it('handleClose called when top-right button clicked', () => {
    const history = createMemoryHistory();
    const component = mount(<UploadToscaInstantiation history={ history }/>)
    const logSpy = jest.spyOn(console, 'log');


    act(() => {
      component.find('[size="lg"]').get(0).props.onHide();
      expect(logSpy).toHaveBeenCalledWith('handleClose called');
    });
  });

  it('fileUploadHandler called when uploading a defined file', () => {
    const component = mount(<UploadToscaInstantiation/>)
    const logSpy = jest.spyOn(console, 'log');

    const event = {
      preventDefault() {
      },
      currentTarget: { files: [file] }
    };

    act(() => {
      component.find('[type="file"]').get(0).props.onChange(event);
      expect(logSpy).toHaveBeenCalledWith('fileUploadHandler called');
      expect(logSpy).toHaveBeenCalledWith('fileDefined called');
    });
  });

  it('fileUploadHandler called when uploading a undefined file', () => {
    const component = mount(<UploadToscaInstantiation/>)
    const logSpy = jest.spyOn(console, 'log');

    const event = {
      preventDefault() {
      },
      currentTarget: { files: [] }
    };

    act(() => {
      component.find('[type="file"]').get(0).props.onChange(event);
      expect(logSpy).toHaveBeenCalledWith('fileUndefined called');
    });
  });
});