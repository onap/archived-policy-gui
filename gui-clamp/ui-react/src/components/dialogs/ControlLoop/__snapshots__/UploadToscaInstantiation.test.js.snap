// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify UploadToscaInstantiation renders correctly 1`] = `
<Styled(Modal)
  backdrop="static"
  keyboard={false}
  onHide={[Function]}
  show={true}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle>
      Tosca Control Loop - Create Instantiation
    </ModalTitle>
  </ModalHeader>
  <div
    style={
      Object {
        "padding": "5px 5px 0px 5px",
      }
    }
  >
    <ModalBody>
      <Form
        inline={false}
        style={
          Object {
            "paddingTop": "20px",
          }
        }
      >
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "render": [Function],
            }
          }
        >
          <FormFile
            accept=".yaml,.yml,.json"
            className="custom-file-label"
            custom={true}
            id="inputGroupFile01"
            label="Please select a file"
            onChange={[Function]}
            type="file"
          />
          <UploadToscaInstantiationFile
            jsonObject={Array []}
            onResponseReceived={[Function]}
          />
          <FormText>
            Only .yaml, .yml and .json files are supported
          </FormText>
        </FormGroup>
        <FormGroup
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "noGutters": false,
              },
              "render": [Function],
            }
          }
        >
          <styled.div />
        </FormGroup>
      </Form>
    </ModalBody>
  </div>
  <ModalFooter>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="null"
      variant="secondary"
    >
      Close
    </Button>
  </ModalFooter>
</Styled(Modal)>
`;
